paymentsheet/src/test/java/com/stripe/android/paymentsheet/paymentdatacollection/bacs/BacsMandateConfirmationViewModelTest.kt

```diff
@@ -0,0 +1,86 @@
+package com.stripe.android.paymentsheet.paymentdatacollection.bacs
+
+import app.cash.turbine.test
+import com.google.common.truth.Truth.assertThat
+import com.stripe.android.core.strings.resolvableString
+import com.stripe.android.paymentsheet.R
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.test.UnconfinedTestDispatcher
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.runTest
+import kotlinx.coroutines.test.setMain
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+
+class BacsMandateConfirmationViewModelTest {
+    @Before
+    fun setup() {
+        Dispatchers.setMain(UnconfinedTestDispatcher())
```

Comment:
Are these needed? If so, how come?

Reply:
The actions are passed through a `SharedFlow` which requires some coroutine scope in order to emit values (can do `tryEmit` synchronously but does not guarantee execution). The `viewModelScope` uses the `MainDispatcher` internally which needs to be overwritten in tests. Otherwise we get this error:
```
Module with the Main dispatcher had failed to initialize. For tests Dispatchers.setMain from kotlinx-coroutines-test module can be used
```
We have a test rule for this called `CoroutineTestRule` that sits in the `stripe-test-e2e` which could be useful elsewhere for other flow tests.

Reply:
Ahh, makes sense. Thanks!